### PR TITLE
Cleanup azure settings usage

### LINKED ISSUE #86: CreationException when elasticsearch.yml is incorrect
I am trying to get Elasticsearch working on an Azure VM with the Azure Elasticsearch plugin. I have followed the steps on the project page right up to the point where I start the service. When I try to start the service I get the following error:

> ...
> [2015-05-16 18:57:36,836][ERROR][bootstrap ] Exception org.elasticsearch.common.inject.CreationException: Guice creation errors:

> 1) No implementation for org.elasticsearch.discovery.Discovery was bound. while locating org.elasticsearch.discovery.Discovery for parameter 2 at org.elasticsearch.discovery.DiscoveryService.(Unknown Source)
at unknown

> 2) No implementation for org.elasticsearch.discovery.Discovery was bound. while locating org.elasticsearch.discovery.Discovery for parameter 3 at org.elasticsearch.node.service.NodeService.(Unknown Source) at unknown

I'm not sure if this is a bug or an error on my part, but if it's human error it would be nice at the very least to have a more descriptive message.

### PR DESCRIPTION
**Pull request by dadoonet**
This commit simplifies and checks azure discovery settings.

Previously, when a user was entering bad azure configuration in its `elasticsearch.yml`, elasticsearch was complaining about Guice injection errors with cryptic messages.

With this commit, when a user starts elasticsearch with wrong settings, `discovery-azure` plugin will fail and will **block starting elasticsearch** with a more obvious message!

```
[2015-12-03 10:03:55,539][INFO ][org.elasticsearch.node   ] [Sean Cassidy] version[3.0.0-SNAPSHOT], pid[38249], build[Unknown/Unknown]
[2015-12-03 10:03:55,540][INFO ][org.elasticsearch.node   ] [Sean Cassidy] initializing ...
[2015-12-03 10:03:55,551][INFO ][org.elasticsearch.plugins] [Sean Cassidy] loaded [discovery-azure], sites []
[2015-12-03 10:03:55,555][WARN ][org.elasticsearch.plugins] [Sean Cassidy] Plugin: discovery-azure implementing onModule by the type is not of Module type class org.elasticsearch.index.IndexModule
[2015-12-03 10:03:55,662][INFO ][org.elasticsearch.env    ] [Sean Cassidy] using [1] data paths, mounts [[/ (/dev/disk1)]], net usable_space [20.7gb], net total_space [464.7gb], spins? [unknown], types [hfs]
[2015-12-03 10:03:55,981][ERROR][org.elasticsearch.plugin.discovery.azure] one or more azure discovery settings are missing. Check elasticsearch.yml file. Should have [cloud.azure.management.subscription.id], [cloud.azure.management.cloud.service.name], [cloud.azure.management.keystore.path] and [cloud.azure.management.keystore.password].
Exception in thread "main" ElasticsearchException[one or more azure discovery settings are missing. Check elasticsearch.yml file. Should have [cloud.azure.management.subscription.id], [cloud.azure.management.cloud.service.name], [cloud.azure.management.keystore.path] and [cloud.azure.management.keystore.password].]
    at org.elasticsearch.plugin.discovery.azure.AzureDiscoveryPlugin.nodeModules(AzureDiscoveryPlugin.java:78)
    at org.elasticsearch.plugins.PluginsService.nodeModules(PluginsService.java:233)
    at org.elasticsearch.node.Node.<init>(Node.java:172)
    ...
```

Also, because we have split `cloud-azure` plugin, we know that this `discovery-azure` plugin is used **only** for discovery purpose on Azure platform. So we can detect easily that a user is embedding a plugin which is not used. And we can now warn him:

```
[2015-12-03 10:13:22,385][INFO ][org.elasticsearch.node   ] [Gregor Shapanka] version[3.0.0-SNAPSHOT], pid[38632], build[Unknown/Unknown]
[2015-12-03 10:13:22,386][INFO ][org.elasticsearch.node   ] [Gregor Shapanka] initializing ...
[2015-12-03 10:13:22,395][INFO ][org.elasticsearch.plugin.discovery.azure] You added discovery-azure plugin but not using it. Check your config or remove it.
[2015-12-03 10:13:22,413][INFO ][org.elasticsearch.plugins] [Gregor Shapanka] loaded [discovery-azure], sites []
```

Sadly, azure Java SDK is swallowing some exceptions and just log with its own logger when something wrong is happening.

So we can't really detect efficiently when the Azure client is created that something is wrong.
We only detect when we start using the client!

This commit immediately checks after the client is created that the client is running fine.
This allows us to also check that at least one service exists for our account and that the service we are supposed to join exists in Azure and fail to start Elasticsearch if it's not the case (see below).

``` java
HostedServiceListResponse list = computeManagementClient.getHostedServicesOperations().list();
if (list.getHostedServices().isEmpty()) {
    throw new ElasticsearchException("no hosted service seems to be available for your azure account.");
}
boolean serviceFound = false;
for (HostedServiceListResponse.HostedService hostedService : list) {
    if (hostedService.getServiceName().equals(serviceName)) {
        serviceFound = true;
    }
}
```

If [Azure issue 565](https://github.com/Azure/azure-sdk-for-java/issues/565) get fixed, we will be able to fail with better messages such as:
- pkstore file does not exist or is not readable
- password for pkstore is incorrect
- pkstore format is incorrect
- ...

If the service name is unknown, elasticsearch will now fail to start with a clear explanation:

```
Caused by: ElasticsearchException[[dadoonet] does not exist for your azure account. To list available services, run in debug mode or check your azure console at https://portal.azure.com/.]
```

The user can run in `debug` mode and get the list of known services:

```
[2015-12-03 14:26:32,435][TRACE][org.elasticsearch.cloud.azure.management] [Seamus Mellencamp] creating new Azure client for [account id here], [dadoonet]
[2015-12-03 14:26:33,292][DEBUG][org.elasticsearch.cloud.azure.management] [Seamus Mellencamp] checking azure client...
[2015-12-03 14:26:34,878][DEBUG][org.elasticsearch.cloud.azure.management] [Seamus Mellencamp] Known hosted services for your azure account:
[2015-12-03 14:26:34,879][DEBUG][org.elasticsearch.cloud.azure.management] [Seamus Mellencamp] - dadoonet-node1
[2015-12-03 14:26:34,879][DEBUG][org.elasticsearch.cloud.azure.management] [Seamus Mellencamp] - dadoonet-tests
```

Closes https://github.com/elastic/elasticsearch-cloud-azure/issues/86.

### CODE REVIEW

**Comment by dadoonet** [comment]
@imotov Would you please review this one please? :D

File: plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeServiceImpl.java [hunk 1]
```diff
@@ -97,16 +89,62 @@ public HostedServiceGetDetailedResponse getServiceDetails() {
         }
     }
 
+    // This method is never called implicitly!
```
Comment (imotov) [inline plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeServiceImpl.java]: If I understand the code correctly, you are not following standard lifecycle for this component. So, maybe it would make sense not to use `AbstractLifecycleComponent` here and instead just have an `initialize` method that would perform initialization. I found these comments are a bit confusing.

File: plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeServiceImpl.java [hunk 2]
```diff
@@ -97,16 +89,62 @@ public HostedServiceGetDetailedResponse getServiceDetails() {
...
     @Override
     protected void doStop() throws ElasticsearchException {
     }
 
+    // This method is never called implicitly!
     @Override
     protected void doClose() throws ElasticsearchException {
+        // This code is never executed!
```
Comment (imotov) [inline plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeServiceImpl.java]: If this code is never executed why does it exist?

**Comment by imotov** [comment]
I understand that this is a hack to work around guice and azure issues, but I am still concerned about removal of cleanup code. So, I am wondering if we can make it a bit less hacky by moving code from `doStart()` to some function that will be called from `configure()`, this function will do all initialization and then return `computeManagementClient` that you can store in a singleton and inject to  `AzureComputeServiceImpl`. This way `AzureComputeServiceImpl` can keep it lifecycle logic and do necessary cleanup in `close()`. What do you think?

File: plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeServiceImpl.java (hunk 2 above)
Comment (dadoonet) [inline plugins/discovery-azure/src/main/java/org/elasticsearch/cloud/azure/management/AzureComputeServiceImpl.java]: Well. I guess I added there 2 years ago before knowing that it won't be ever executed! 
I did not remove it as I think we should may be find a better solution when the node stops to actually close the azure client.
I think this is not related to this PR though. But I should probably add a TODO in my comment.

**Comment by dadoonet** [comment]
I'll give a try to your suggestion. Thanks for the comments so far. Reassigning the PR to me now.
I'll ping you when updated.

**Comment by dadoonet** [comment]
@imotov I rebased on master and added a new commit.

I still don't see from which class I should call `close()` method. Any idea?

**Comment by imotov** [comment]
> I still don't see from which class I should call close() method. Any idea?

@dadoonet I tried to explain in my previous comment but I probably wasn't very clear. Could you ping me when you have time and we can talk about it? It might be easier to explain in person.

**Comment by clintongormley** [comment]
@dadoonet are you coming back to this one?

**Comment by dadoonet** [comment]
I tried recently to just merge changes from master in my branch but that was not as easy as I'd expect.
I definitely want to do it so please let's keep it opened and assigned to me.

**Comment by dadoonet** [comment]
As we deprecated azure-classic in #19186, I think this PR is pretty much useless now.
I'm closing it.

Thanks @imotov for the review though!
